### PR TITLE
Additional externs for ecdsa recovery and keccak

### PR DESCRIPTION
**Pull request by gavofyork**
Needed for Poilkadot's claims module

### CODE REVIEW

File: core/executor/src/wasm_executor.rs [hunk 1]
```diff
@@ -461,6 +463,15 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
 		this.memory.set(out, &result).map_err(|_| UserError("Invalid attempt to set result in ext_blake2_256"))?;
 		Ok(())
 	},
+	ext_keccak_256(data: *const u8, len: u32, out: *mut u8) => {
+		let result = if len == 0 {
+			tiny_keccak::keccak256(&[0u8; 0])
+		} else {
+			tiny_keccak::keccak256(&this.memory.get(data, len as usize).map_err(|_| UserError("Invalid attempt to get data in ext_blake2_256"))?)
```
Comment (pepyakin) [inline core/executor/src/wasm_executor.rs]:
```suggestion
			tiny_keccak::keccak256(&this.memory.get(data, len as usize).map_err(|_| UserError("Invalid attempt to get data in ext_keccak_256"))?)
```

File: core/executor/src/wasm_executor.rs [hunk 2]
```diff
@@ -461,6 +463,15 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
...
+		};
+		this.memory.set(out, &result).map_err(|_| UserError("Invalid attempt to set result in ext_blake2_256"))?;
```
Comment (pepyakin) [inline core/executor/src/wasm_executor.rs]:
```suggestion
		this.memory.set(out, &result).map_err(|_| UserError("Invalid attempt to set result in ext_keccak_256"))?;
```

File: core/executor/src/wasm_executor.rs [hunk 3]
```diff
@@ -461,6 +463,15 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
 		this.memory.set(out, &result).map_err(|_| UserError("Invalid attempt to set result in ext_blake2_256"))?;
 		Ok(())
 	},
+	ext_keccak_256(data: *const u8, len: u32, out: *mut u8) => {
+		let result = if len == 0 {
```
Comment (pepyakin) [inline core/executor/src/wasm_executor.rs]: For documentation purposes, consider specifying the type `[u8; 32]` explicitly.

File: core/sr-io/with_std.rs [hunk 4]
```diff
@@ -200,6 +204,18 @@ pub fn ed25519_verify<P: AsRef<[u8]>>(sig: &[u8; 64], msg: &[u8], pubkey: P) ->
 	ed25519::verify(sig, msg, pubkey)
 }
 
+/// Verify and recover a SECP256k1 ECDSA signature.
+/// - `sig` is passed in RSV format. V should be either 0/1 or 27/28.
+/// - returns `None` if the signatue is bad, the 64-byte pubkey (doesn't include the 0x04 prefix).
```
Comment (pepyakin) [inline core/sr-io/with_std.rs]: It can't return `None`, can it?

**Comment by pepyakin** [comment]
Please resolve merge conflict in Cargo.lock

https://github.com/paritytech/substrate/pull/1583/files#diff-d2ede298d9a75c849170d6ef285eda1eR4848

File: core/executor/src/wasm_executor.rs (hunk 3 above)
Comment (gavofyork) [reply core/executor/src/wasm_executor.rs]: Not sure what you mean here..

Comment (pepyakin) [reply core/executor/src/wasm_executor.rs]: Basically this

```suggestion
		let result: [u8; 32] = if len == 0 {
```

it allows the reader of the code realize that only `out[0..32]` bytes are written.

Comment (gavofyork) [reply core/executor/src/wasm_executor.rs]: Well.. that was just copy-pasted code, so to keep it consistent with the other hash impls I'd need to touch more than this PR is meant to...

**Pull request merged**